### PR TITLE
common/hobject: a minor fix and performance gain to hobjects listing

### PR DESCRIPTION
**Pull request by ronen-fr**
The fix: the 'char >= 127' test was always(*) false.
    (*) ostensibly - compiler dependent

Speedup:     
   Using charconv::to_chars().
    Anecdotal speedup handling an assorted mix of strings: 2.4X
    (tested on my laptop)
 
Signed-off-by: Ronen Friedman <rfriedma@redhat.com>
Reviewed-by: Kefu Chai <kchai@redhat.com>
Reviewed-by: Josh Durgin <jdurgin@redhat.com>

### CODE REVIEW

**Comment by ronen-fr** [comment]
smoke test suite OK

File: src/common/hobject.cc [hunk 1]
```diff
@@ -210,15 +212,18 @@ void hobject_t::generate_test_instances(list<hobject_t*>& o)
...
 {
-  for (string::const_iterator i = in.begin(); i != in.end(); ++i) {
-    int k = (int)(unsigned char)(*i);
-    if (k == '%' || k == ':' || k == '/' || k < 32 || k >= 127) {
-      out->push_back('%');
-      char buf[3];
-      snprintf(buf, sizeof(buf), "%02x", (int)(unsigned char)k);
+  for (auto ii = in.begin(); ii != in.end(); ++ii) {
+    int i = (int)(unsigned char)(*ii);
+    if (i < 0x0f) {
```
Comment (jdurgin) [inline src/common/hobject.cc:217]: should this be <= ?

File: src/common/hobject.cc [hunk 2]
```diff
@@ -210,14 +212,18 @@ void hobject_t::generate_test_instances(list<hobject_t*>& o)
...
-  for (string::const_iterator i = in.begin(); i != in.end(); ++i) {
-    if (*i == '%' || *i == ':' || *i == '/' || *i < 32 || *i >= 127) {
-      out->push_back('%');
-      char buf[3];
-      snprintf(buf, sizeof(buf), "%02x", (int)(unsigned char)*i);
+  for (auto ii = in.begin(); ii != in.end(); ++ii) {
+    int i = (int)(unsigned char)(*ii);
+    if (i < 0x0f) {
+      char buf[4] = {'%', '0'};
+      to_chars(buf+2, buf+3, i, 16);
```
Comment (tchaikov) [inline src/common/hobject.cc:219]:
```suggestion
      std::to_chars(buf + 2, buf + 3, i, 16);
```

File: src/common/hobject.cc (hunk 1 above)
Comment (ronen-fr) [reply src/common/hobject.cc:217]: Indeed. A last minute bug

**Comment by ronen-fr** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
test failure tracked by https://tracker.ceph.com/issues/48850

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by ronen-fr** [comment]
(latest forced-push: a missing std::)

File: src/common/hobject.cc [hunk 3]
```diff
@@ -210,14 +212,18 @@ void hobject_t::generate_test_instances(list<hobject_t*>& o)
 
 static void append_out_escaped(const string &in, string *out)
 {
-  for (string::const_iterator i = in.begin(); i != in.end(); ++i) {
-    if (*i == '%' || *i == ':' || *i == '/' || *i < 32 || *i >= 127) {
-      out->push_back('%');
-      char buf[3];
-      snprintf(buf, sizeof(buf), "%02x", (int)(unsigned char)*i);
+  for (auto ii = in.begin(); ii != in.end(); ++ii) {
```
Comment (tchaikov) [inline src/common/hobject.cc:215]:
```suggestion
  for (auto c : in) {
```

File: src/common/hobject.cc [hunk 4]
```diff
@@ -210,14 +212,18 @@ void hobject_t::generate_test_instances(list<hobject_t*>& o)
...
+      char buf[4] = {'%', '0'};
+      std::to_chars(buf + 2, buf + 3, i, 16);
+      out->append(buf);
+    } else if (i >= 127 || i == '%' || i == ':' || i == '/') {
+      char buf[4] = {'%'};
+      std::to_chars(buf + 1, buf + 3, i, 16);
       out->append(buf);
     } else {
-      out->push_back(*i);
+      out->push_back(i);
```
Comment (tchaikov) [inline src/common/hobject.cc:226]:
```suggestion
      out->push_back(c);
```

**Review by tchaikov — APPROVED** [review]

File: src/common/hobject.cc [hunk 5]
```diff
@@ -1,6 +1,8 @@
 // -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
 // vim: ts=8 sw=2 smarttab
 
+#include <charconv>
```
Comment (tchaikov) [inline src/common/hobject.cc:4]: in general, we don't add "Reviewed-by" in non-merge commits.

File: src/common/hobject.cc [hunk 6]
```diff
@@ -210,14 +212,18 @@ void hobject_t::generate_test_instances(list<hobject_t*>& o)
...
 static void append_out_escaped(const string &in, string *out)
 {
-  for (string::const_iterator i = in.begin(); i != in.end(); ++i) {
-    if (*i == '%' || *i == ':' || *i == '/' || *i < 32 || *i >= 127) {
-      out->push_back('%');
-      char buf[3];
-      snprintf(buf, sizeof(buf), "%02x", (int)(unsigned char)*i);
+  for (auto c : in) {
+    int i = (int)(unsigned char)(c);
+    if (i <= 0x0f) {
```
Comment (rzarzynski) [inline src/common/hobject.cc:217]: Why `<= 0x0f` when previously it was `< 32`?

Comment (ronen-fr) [reply src/common/hobject.cc:217]: Just what happens when one resolves a review comment too late at night.
Now it's correct: '<=0x0f' handled separately from the rest of the <32

File: src/common/hobject.cc [hunk 7]
```diff
@@ -210,14 +212,18 @@ void hobject_t::generate_test_instances(list<hobject_t*>& o)
...
+      char buf[4] = {'%', '0'};
+      std::to_chars(buf + 2, buf + 3, i, 16);
+      out->append(buf);
+    } else if (i < 32 || i >= 127 || i == '%' || i == ':' || i == '/') {
```
Comment (rzarzynski) [inline src/common/hobject.cc:221]: ACK.

**Review by rzarzynski — APPROVED** [review]

**Comment by tchaikov** [comment]
https://pulpito.ceph.com/kchai-2021-07-13_00:59:36-rados-wip-kefu-testing-2021-07-12-2312-distro-basic-smithi/

**Pull request merged**